Fix validation issues

steveyken opened the pull request:
The following features have been implemented:

- all invalid core and custom fields should have red background highlighting
- required custom fields are now indicated with '*'
- field groups containing validation errors should always be expanded when returning the error messages (see screenshot)
- FIX contact first_name showing '*' even when not required
- added html5 'in browser' validation to core fields
- upon form submit, reveal any hidden invalid fields in order to facilitate correction (otherwise, user has no idea why the form won't submit) 
- FIX remove '*' from several fields in task that looked required but have no validation

![image](https://github.com/fatfreecrm/fat_free_crm/assets/149198/d407ce83-90df-464c-aa2e-7a65c0fa96fc)

steveyken commented:
Hold for now. Just found a bug when creating a new Contact... the above works for 'edit' but not for 'new'

steveyken commented:
This is now ready for actual merge again. Assuming branch checks pass.

CloCkWeRX commented:
with ..

remove html5 'in browser' validation from custom fields to enable Rails validations to provide a better experience (form doesn't submit when section containing invalid custom field is collapsed and therefore not in view)


In general I really like to keep these behaviours. Better mobile UX, ability to use patterns, titles, etc.

The problem with collapsed/display none elements that are required is usually fixed by one of:
- marking the elements disabled when they are hidden, so that the browser knows they aren't required. Use case is one form answer makes a subsequent area of the form mandatory
- handling the before submit / validation event and when invalid, hidden elements are detected, marking the whole section visually invalid or expanding it

It's more work, but you get the best of both worlds then - instant UX feedback for the basics, and higher level domain errors from rails after submit.

CloCkWeRX reviewed (approved):
Well, I wrote all that but don't actually see where you suppressed HTML5 behaviours, so lgtm

steveyken commented:
Yes, that all makes sense. My js isn't so strong but I'll give it a go. We'd need to expand sections that have invalid fields. I could potentially add html5 'required' attribute to the core fields that require validation too if you like e.g. account name, or contact first_name / last_name (dependent on settings).

Will update this PR in a bit.

p.s. the html5 validation was removed in app/models/fields/field.rb where 'required' was taken out of input_options which is fed to simple form in the view templates.

steveyken commented:
I've updated the code now to re-enable html5 validations for custom fields and also to extend it to core fields also. If a custom field is hidden and invalid, some form submit JS will now open the section that contains the invalid field in order to display the validation message.

Bonus: removed '*' required asterisk from a few fields in task that had no underlying validation requirements.

I believe this is ready to merge now.

steveyken commented:
Ok for me to merge?

CloCkWeRX commented:
Sorry distracted by shiny things, let me look

CloCkWeRX commented:
That's really nice. Thank you for going the extra mile

steveyken commented:
👍
